fs/remote/resolver.go

```diff
@@ -430,114 +273,104 @@ func (f *httpFetcher) fetch(ctx context.Context, rs []region, retry bool) (multi
+		if retry {
+			log.G(ctx).Infof("Received status code: %v. Refreshing URL and retrying...", res.Status)
+			if err := f.refreshURL(ctx); err != nil {
+				return nil, fmt.Errorf("%w: status %v: %w", ErrFailedToRefreshURL, res.Status, err)
+			}
+			return f.fetch(ctx, rs, false)
 		}
-		return f.fetch(ctx, rs, false)
-	} else if retry && res.StatusCode == http.StatusBadRequest && !singleRangeMode {
-		log.G(ctx).Infof("Received status code: %v. Setting single range mode and retrying...", res.Status)
-
+	case http.StatusBadRequest:
 		// gcr.io (https://storage.googleapis.com) returns 400 on multi-range request (2020 #81)
-		f.singleRangeMode()            // fallbacks to singe range request mode
-		return f.fetch(ctx, rs, false) // retries with the single range mode
+		if retry && !singleRangeMode {
+			log.G(ctx).Infof("Received status code: %v. Setting single range mode and retrying...", res.Status)
+			// fallback and retry with  range request mode
+			f.singleRangeMode()
+			return f.fetch(ctx, rs, false)
+		}
 	}
-
-	return nil, fmt.Errorf("unexpected status code: %v", res.Status)
+	return nil, fmt.Errorf("%w on fetch: %v", ErrUnexpectedStatusCode, res.Status)
 }
 
 func (f *httpFetcher) check() error {
 	ctx := context.Background()
-	if f.timeout > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, f.timeout)
-		defer cancel()
-	}
 	f.urlMu.Lock()
-	url := f.url
+	url := f.realBlobURL
 	f.urlMu.Unlock()
 	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
 	if err != nil {
-		return fmt.Errorf("check failed: failed to make request: %w", err)
+		return fmt.Errorf("check failed: %w", err)
 	}
-	req.Close = false
 	req.Header.Set("Range", "bytes=0-1")
-	res, err := f.tr.RoundTrip(req)
+	res, err := f.roundTripper.RoundTrip(req)
 	if err != nil {
-		return fmt.Errorf("check failed: failed to request to registry: %w", err)
+		return fmt.Errorf("check failed: %w: %w", ErrRequestFailed, err)
 	}
-	defer func() {
-		io.Copy(io.Discard, res.Body)
-		res.Body.Close()
-	}()
+	defer socihttp.Drain(res.Body)
 	if res.StatusCode == http.StatusOK || res.StatusCode == http.StatusPartialContent {
 		return nil
 	} else if res.StatusCode == http.StatusForbidden {
 		// Try to re-redirect this blob
 		rCtx := context.Background()
-		if f.timeout > 0 {
-			var rCancel context.CancelFunc
-			rCtx, rCancel = context.WithTimeout(rCtx, f.timeout)
-			defer rCancel()
-		}
 		if err := f.refreshURL(rCtx); err == nil {
 			return nil
 		}
-		return fmt.Errorf("failed to refresh URL on status %v", res.Status)
+		return fmt.Errorf("%w: status %v", ErrFailedToRefreshURL, res.Status)
 	}
 
-	return fmt.Errorf("unexpected status code %v", res.StatusCode)
+	return fmt.Errorf("%w on check: %v", ErrUnexpectedStatusCode, res.StatusCode)
 }
 
 func (f *httpFetcher) refreshURL(ctx context.Context) error {
-	newURL, err := redirect(ctx, f.blobURL, f.tr, f.timeout)
+	newRealBlobURL, err := redirect(ctx, f.baseBlobURL, f.roundTripper)
```

Comment:
> Hmm, I wonder if we should have some instrumentation/metrics around the number of in-flight requests to the registry per layer (as a follow up).

Yeah, there are some other metrics we should probably track as-well, like timeout count. I'll create an issue for adding some more HTTP metrics and attach to the tracking issue #904.

> It's not new in your changes, but I'm realizing that every outstanding request using the old url will refresh the URL, not just the first. That's probably OK if the number of requests for a specific layer is small.

Actually, would it make sense to lock the URL over the entire refresh? The downside/worst case would be some sort of head of line blocking where if the current refresh errors out after expiring all of it's retries then it will have a cascading affect on all the other waiting span requests. The upside is that we only make a single request to refresh the blob on expiry.

Reply:
This wouldn't _fix_  the problem, I suppose it would just make it harder to hit.

Reply:
I'll open another issue for this.